bpf-gpl/tc.c

```diff
@@ -1305,6 +1305,38 @@ static CALI_BPF_INLINE struct fwd calico_tc_skb_accepted(struct __sk_buff *skb,
 	goto deny;
 
 nat_encap:
+	/* We are about to encap return trafic that originated on the local host
+	 * namespace - host networked pod. Routing was based on the dst IP of
```

Comment:
Here, when you mention routing, do you mean the routing choice from this machine back to the machine that forwarded us this connection?

Does "dst IP of the original client" mean the destination IP address to which the client (e.g. an NLB listener) connected, or does that mean that from our perspective here our destination address for return traffic is the client's IP address?

Reply:
> Here, when you mention routing, do you mean the routing choice from this machine back to the machine that forwarded us this connection?

Routing of the packet we are just processing. The packet was routed based on the dst IP, which was the original client's IP at that time. Hence the link layer picked a wrong MAC.

> Does "dst IP of the original client" mean the destination IP address to which the client (e.g. an NLB listener) connected, or does that mean that from our perspective here our destination address for return traffic is the client's IP address?

original client is whoever tried to reach the nodeport. It may be an LB, it may be an external client if the nodeport was exposed to the outter world.